File: matlab/RadialBasisFunction.m
% This file is part of the SPLINTER library.
% Copyright (C) 2012 Bjarne Grimstad (bjarne.grimstad@gmail.com).
%
% This Source Code Form is subject to the terms of the Mozilla Public
% License, v. 2.0. If a copy of the MPL was not distributed with this
% file, You can obtain one at http://mozilla.org/MPL/2.0/.

classdef RadialBasisFunction < Approximant
    properties (Access = protected)
        Handle
        
        Constructor_function = 'rbf_init';
    end

    methods
        function obj = RadialBasisFunction(dataTable, type, normalized)
            % Set to -1 so we don't try to delete the library instance in case type is invalid
            obj.Handle = -1;
            
            % Default to Thin plate spline
            if(~exist('type', 'var'))
                type_index = 1;
            else
                switch type
                    case RadialBasisFunctionType.Thin_plate_spline
                        type_index = 1;
                    case RadialBasisFunctionType.Multiquadric
                        type_index = 2;
                    case RadialBasisFunctionType.Inverse_quadric
                        type_index = 3;
                    case RadialBasisFunctionType.Inverse_multiquadric
                        type_index = 4;
                    case RadialBasisFunctionType.Gaussian
                        type_index = 4;
                end 
            end
            
            if(~exist('normalized', 'var'))
                normalized = 0;
            end
            
            % Make sure that normalized is 0 or 1
            if(normalized ~= 0)
                normalized = 1;
            else
                normalized = 0;
            end
            
            obj.Handle = Splinter.getInstance().call(obj.Constructor_function, dataTable.get_handle(), type_index,  normalized);
        end
    end
end

File: python/radialbasisfunction.py
# This file is part of the SPLINTER library.
# Copyright (C) 2012 Bjarne Grimstad (bjarne.grimstad@gmail.com).
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.


import sys
from approximant import Approximant
from utilities import *


class RBFType:
	THIN_PLATE_SPLINE = 1
	MULTIQUADRIC = 2
	INVERSE_QUADRIC = 3
	INVERSE_MULTIQUADRIC = 4
	GAUSSIAN = 5

class RadialBasisFunction(Approximant):
	def __init__(self, dataTableOrFileName, rbfType=None):
		# If string we load the PolynomialRegression from the file
		if isString(dataTableOrFileName):
			fileName = getCString(dataTableOrFileName)
			self._handle = SPLINTER.call(SPLINTER.getHandle().rbf_load_init, fileName)
		
		# Else, construct PolynomialRegression from DataTable
		else:
			if rbfType is None:
				rbfType = RBFType.THIN_PLATE_SPLINE
			if rbfType < 1 or rbfType > 5:
				raise Exception("Error: Invalid rbfType: " + str(rbfType) + ". For valid values see the class RBFType.")
			
			dataTable = dataTableOrFileName
			self._handle = SPLINTER.call(SPLINTER.getHandle().rbf_init, dataTable.getHandle(), rbfType)
		
		
if __name__ == "__main__":
	import SPLINTER
	SPLINTER.load("/home/anders/SPLINTER/build/release/libsplinter-matlab-1-4.so")
	
	
	from datatable import DataTable
	
	def f(x):
		return x[0]*x[1]
	
	d = DataTable()
	for i in range(10):
		for j in range(10):
			d.addSample([i,j], f([i,j]))
	
	b = RadialBasisFunction(d, RBFType.GAUSSIAN)
	for i in range(10):
		for j in range(10):
			print(str(b.eval([0.9*i,0.9*j])) + " == " + str(0.81*i*j))
	
	print(b.evalJacobian([3,3]))
	print(b.evalHessian([3,3]))
	
	b.save("test.rbf")
	b2 = RadialBasisFunction("test.rbf")
	
	print(b.eval([2,3]))
	print(b2.eval([2,3]))